crates/ty_python_semantic/src/types.rs

```diff
@@ -6802,6 +6813,51 @@ impl<'db> TypeVarBoundOrConstraints<'db> {
     }
 }
 
+/// Error returned if a type is not awaitable.
+#[derive(Debug)]
+enum AwaitError<'db> {
+    /// `__await__` is either missing, potentially unbound or cannot be called with provided
+    /// arguments.
+    Call(CallDunderError<'db>),
+    /// `__await__` resolved successfully, but its return type is known not to be a generator.
+    InvalidReturnType(Type<'db>),
+}
+
+impl<'db> AwaitError<'db> {
+    fn report_diagnostic(
+        &self,
+        context: &InferContext<'db, '_>,
+        context_expression_type: Type<'db>,
+        content_expression_node: ast::AnyNodeRef,
+    ) {
+        let Some(builder) = context.report_lint(&INVALID_AWAIT, content_expression_node) else {
+            return;
+        };
+
+        let db = context.db();
+
+        let mut diag = builder.into_diagnostic(
+            format_args!("`{type}` is not awaitable", type = context_expression_type.display(db)),
+        );
+        match self {
+            Self::Call(CallDunderError::CallError(..)) => {
+                // TODO: Indicate definition error, not usage error.
+                diag.info(
+                    "`__await__` expects additional arguments and cannot be called implicitly",
+                );
+            }
+            Self::Call(CallDunderError::PossiblyUnbound(..)) => {
+                diag.info("`__await__` is possibly unbound");
+            }
+            Self::Call(CallDunderError::MethodNotAvailable) => diag.info("`__await__` is missing"),
+            Self::InvalidReturnType(return_type) => diag.info(format_args!(
+                "`__await__` returns `{return_type}`, which is not a valid iterator",
+                return_type = return_type.display(db)
+            )),
+        }
+    }
```

Comment:
Is it possible that in these cases we can highlight the range of the `__await__` method, or of the definition of the type that is lacking an `__await__` method?